security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,141 @@
+// limitations under the License.
+
+package custom
+
+import (
+	"context"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"google.golang.org/grpc/metadata"
+
+	mesh "istio.io/api/mesh/v1alpha1"
+	pb "istio.io/api/security/v1alpha1"
+	"istio.io/istio/security/pkg/pki/signingapi"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultRequestTimeout = 10 * time.Second
+	bearerTokenPrefix     = "Bearer "
+)
+
+var (
+	cLog = log.RegisterScope("CustomCAClient", "Custom CA Integration Log", 0)
+
+	// ClientCertDirENV Location of a local or mounted Custom CA Client TLS Certificates
+	// Includes: root-cert.pem, client.crt, client.key
+	// Users require to create kubernetes secret and mount them into this location
+	ClientCertDirENV = env.RegisterStringVar("CUSTOM_CA_CERT_DIR", "./var/lib/istio/custom-ca-tls",
+		"Location of a local or mounted Custom CA TLS certificates")
+)
+
+// CAClient generates keys and certificates for Istio identities.
+type CAClient struct {
+	signingAPIClient *signingapi.Client
+	keyCertBundle    util.KeyCertBundle
+	jwtToken         string
+}
+
+// IsUsingCustomCAForWorkloadCertificate return true if users configure CustomCA for signing Workload Certificates
+// TODO: Support for Istio Agent side
+func IsUsingCustomCAForWorkloadCertificate(caOpts *mesh.MeshConfig_CA) bool {
+	return caOpts != nil && caOpts.Address != "" && caOpts.IstiodSide
+}
+
+// NewCAClient returns a new CAClient instance.
+// CustomCAClient will use JWT TOKEN from Istiod to create Authorization header each forwarding request to External CA
+// External CA just need to verify Istiod's identity each forwarding request is sufficient
+func NewCAClient(caOpts *mesh.MeshConfig_CA, keyCertBundle util.KeyCertBundle, jwtToken string) (*CAClient, error) {
+	timeOutRequest, err := types.DurationFromProto(caOpts.GetRequestTimeout())
+	if err != nil {
+		timeOutRequest = defaultRequestTimeout
+	}
+	signingAPIClient := signingapi.New(caOpts.GetAddress(), caOpts.TlsSettings, timeOutRequest)
+	c := &CAClient{
+		keyCertBundle:    keyCertBundle,
+		signingAPIClient: signingAPIClient,
+		jwtToken:         jwtToken,
+	}
+
+	err = c.signingAPIClient.Connect(nil)
+	if err != nil {
+		cLog.Errorf("can not connect to Custom CA Address: %v", err)
+		return nil, fmt.Errorf("can not connect to Custom CA Address: %v", err)
+	}
+	return c, nil
+}
+
+func responseTimeTrack(start time.Time, name string) {
+	elapsed := time.Since(start)
+	cLog.Debugf("%s took %s", name, elapsed)
+}
+
+// CreateCertificate is similar to Sign but returns the leaf cert and the entire cert chain.
+func (c *CAClient) CreateCertificate(ctx context.Context,
+	req *pb.IstioCertificateRequest) (*pb.IstioCertificateResponse, error) {
+	if c.jwtToken != "" {
+		// add Bearer prefix, which is required by Citadel.
+		token := bearerTokenPrefix + c.jwtToken
+		ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("Authorization", token))
```

Comment:
Can you use the standard gRPC method for adding credentials? https://github.com/istio/istio/blob/2347a04bd53ec44c62142cc14b3601b5858aa738/pkg/istio-agent/xds-proxy.go#L357 for an example.